src/MongoDbOperationStore.ts

```diff
@@ -0,0 +1,139 @@
+   * whether the initialization is resuming from a previous stoppage with state
+   * stored in the backend mongodb database; if resuming is false, then a new
+   * database and collection is created.
+   */
+  public async initialize (resuming: boolean): Promise<void> {
+    const databaseName = 'sidetree';
+    const collectionsName = 'operations';
+    this.client = await MongoClient.connect(this.serverUrl);
+    this.db = this.client.db(databaseName);
+
+    // If we are resuming, then the collection, indexes already exist.
+    // Otherwise, we need to create the collection and define an index on
+    // (did, transactionNumber, operationIndex), first dropping the collection if
+    // it exists.
+    if (resuming) {
+      this.collection = this.db.collection(collectionsName);
+    } else {
+      await this.db.collection(collectionsName).drop();
+      this.collection = await this.db.createCollection(collectionsName, { strict: false });
+      await this.collection.createIndex({ didUniqueSuffix: 1, transactionNumber: 1, operationIndex: 1 });
+    }
+  }
+
+  /**
+   * Implement OperationStore.put.
+   */
+  public async put (operation: Operation): Promise<void> {
+    const mongoOperation = this.convertToMongoOperation(operation);
+    await this.collection!.insertOne(mongoOperation);
+  }
+
+  /**
+   * Implement OperationStore.putBatch
+   */
+  public async putBatch (operations: Array<Operation>): Promise<void> {
+    let batch = this.collection!.initializeUnorderedBulkOp();
+
+    for (const operation of operations) {
+      const mongoOperation = this.convertToMongoOperation(operation);
+      batch.insert(mongoOperation);
+    }
+
+    await batch.execute();
+  }
+
+  /**
+   * Get an iterator that returns all operations with a given
+   * didUniqueSuffix ordered by (transactionNumber, operationIndex)
+   * ascending.
+   */
+  public async get (didUniqueSuffix: string): Promise<Iterable<Operation>> {
+    const mongoOperations = await this.collection!.find({ didUniqueSuffix }).sort({ transactionNumber: 1, operationIndex: 1 }).toArray();
+    return mongoOperations.map(this.convertToOperation);
+  }
+
+  /**
+   * Delete all operations with transaction number greater than the
+   * provided parameter.
+   */
+  public async delete (transactionNumber?: number): Promise<void> {
+    if (transactionNumber) {
+      await this.collection!.deleteMany({ transactionNumber: { $gt: transactionNumber } });
+    } else {
+      await this.collection!.deleteMany({});
+    }
+  }
+
+  /**
+   * Convert a Sidetree operation to a more minimal MongoOperation object
+   * that can be stored on MongoDb. The MongoOperation object has sufficient
+   * information to reconstruct the original operation.
+   */
+  private convertToMongoOperation (operation: Operation): MongoOperation {
+    // Convert the operation buffer to string (mongo read-write not handled seamlessly for buffer types)
```

Comment:
Sanity: I think there is actually no need to encode it, the operation itself should be in JSON, so string as is sufficient.

Reply:
That is what I tried first, but ran into problems with storing and retreiving Buffer objects. If you know how to fix it please do so - I agree it is better to avoid convert*() operations